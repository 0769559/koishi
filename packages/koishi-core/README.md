# shiki-core
 
[![npm](https://img.shields.io/npm/v/koishi-core?style=flat-square)](https://www.npmjs.com/package/koishi-core)

A bot based on CoolQ HTTP API.
