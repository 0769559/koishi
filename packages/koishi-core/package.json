{
  "name": "koishi-core",
  "description": "Core features for Koishi",
  "version": "1.3.1",
  "main": "dist/index.js",
  "typings": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "author": "Shigma <1700011071@pku.edu.cn>",
  "license": "MIT",
  "scripts": {
    "jest": "jest tests/.*\\.spec\\.ts --runInBand",
    "lint": "eslint src --ext .ts",
    "precommit": "yarn lint",
    "prepack": "tsc -b"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/koishijs/koishi.git"
  },
  "bugs": {
    "url": "https://github.com/koishijs/koishi/issues"
  },
  "homepage": "https://github.com/koishijs/koishi/tree/master/packages/koishi-core#readme",
  "keywords": [
    "bot",
    "qqbot",
    "cqhttp",
    "coolq",
    "chatbot",
    "koishi"
  ],
  "devDependencies": {
    "@types/debug": "^4.1.5",
    "@types/ws": "^6.0.4",
    "get-port": "^5.1.0",
    "koishi-test-utils": "^1.2.1"
  },
  "dependencies": {
    "axios": "^0.19.1",
    "escape-string-regexp": "^2.0.0",
    "koishi-utils": "^1.0.2",
    "leven": "^3.1.0",
    "ws": "^7.2.1"
  }
}
