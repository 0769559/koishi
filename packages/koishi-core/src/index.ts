export * from './app'
export * from './command'
export * from './context'
export * from './database'
export * from './meta'
export * from './sender'
export * from './server'
export * from './plugins/help'
export * from './plugins/suggest'
export * from './plugins/throttle'
export * from './plugins/queue'
export * from './plugins/validate'

import './plugins/help'
import './plugins/suggest'
import './plugins/throttle'
import './plugins/queue'
import './plugins/validate'
