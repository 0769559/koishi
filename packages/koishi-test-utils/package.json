{
  "name": "koishi-test-utils",
  "description": "Test utilities for Koishi",
  "version": "4.0.0-alpha.39",
  "main": "dist/index.js",
  "typings": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "author": "Shigma <1700011071@pku.edu.cn>",
  "license": "MIT",
  "scripts": {
    "lint": "eslint src --ext .ts",
    "prepack": "tsc -b"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/koishijs/koishi.git"
  },
  "bugs": {
    "url": "https://github.com/koishijs/koishi/issues"
  },
  "homepage": "https://github.com/koishijs/koishi/packages/test-utils#readme",
  "keywords": [
    "bot",
    "qqbot",
    "cqhttp",
    "coolq",
    "chatbot",
    "koishi",
    "test",
    "unit",
    "utilities"
  ],
  "peerDependencies": {
    "jest": "^26.1.0"
  },
  "devDependencies": {
    "@types/debug": "^4.1.5",
    "@types/jest": "^26.0.7"
  },
  "dependencies": {
    "axios": "^0.19.2",
    "get-port": "^5.1.1",
    "koishi-core": "^2.0.0-alpha.39",
    "koishi-utils": "^2.1.2"
  }
}
