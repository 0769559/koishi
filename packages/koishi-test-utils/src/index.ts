import * as memory from './memory'

export { memory }

export * from './app'
export * from './database'
export * from './memory'
export * from './mocks'
export * from './utils'
export * from './server'
export * from './session'

jest.mock('koishi-utils', () => {
  const utils1 = jest.requireActual('koishi-utils')
  const utils2 = jest.requireActual('./koishi')
  return { ...utils1, ...utils2 }
})
