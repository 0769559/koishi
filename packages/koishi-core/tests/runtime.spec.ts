import { MockedApp } from 'koishi-test-utils'
import { messages, Meta } from 'koishi-core'
import { format } from 'util'

const app = new MockedApp()
const session1 = app.createSession('user', 789)
const session2 = app.createSession('group', 456, 321)

app.command('foo <text>', { checkArgCount: true })
  .shortcut('bar1', { args: ['bar'] })
  .shortcut('bar4', { oneArg: true, fuzzy: true })
  .action(({ meta }, bar) => {
    return meta.$send('foo' + bar)
  })

app.command('fooo', { checkUnknown: true, checkRequired: true })
  .shortcut('bar2', { options: { text: 'bar' } })
  .shortcut('bar3', { prefix: true, fuzzy: true })
  .option('-t, --text <bar>')
  .action(({ meta, options }) => {
    return meta.$send('fooo' + options.text)
  })

describe('command prefix', () => {
  beforeAll(() => {
    app.options.similarityCoefficient = 0
  })

  afterAll(() => {
    app.options.commandPrefix = null
    app.options.similarityCoefficient = 0.4
    app.prepare()
  })

  test('no prefix', async () => {
    app.options.commandPrefix = null
    app.prepare()

    await session1.shouldHaveReply('foo bar', 'foobar')
    await session2.shouldHaveReply('foo bar', 'foobar')
    await session1.shouldHaveNoReply('!foo bar')
    await session2.shouldHaveNoReply('!foo bar')
    await session1.shouldHaveNoReply('.foo bar')
    await session2.shouldHaveNoReply('.foo bar')
  })

  test('single prefix', async () => {
    app.options.commandPrefix = '!'
    app.prepare()

    await session1.shouldHaveReply('foo bar', 'foobar')
    await session2.shouldHaveNoReply('foo bar')
    await session1.shouldHaveReply('!foo bar', 'foobar')
    await session2.shouldHaveReply('!foo bar', 'foobar')
    await session1.shouldHaveNoReply('.foo bar')
    await session2.shouldHaveNoReply('.foo bar')
  })

  test('multiple prefixes', async () => {
    app.options.commandPrefix = ['!', '.']
    app.prepare()

    await session1.shouldHaveReply('foo bar', 'foobar')
    await session2.shouldHaveNoReply('foo bar')
    await session1.shouldHaveReply('!foo bar', 'foobar')
    await session2.shouldHaveReply('!foo bar', 'foobar')
    await session1.shouldHaveReply('.foo bar', 'foobar')
    await session2.shouldHaveReply('.foo bar', 'foobar')
  })

  test('optional prefix', async () => {
    app.options.commandPrefix = ['.', '']
    app.prepare()

    await session1.shouldHaveReply('foo bar', 'foobar')
    await session2.shouldHaveReply('foo bar', 'foobar')
    await session1.shouldHaveNoReply('!foo bar')
    await session2.shouldHaveNoReply('!foo bar')
    await session1.shouldHaveReply('.foo bar', 'foobar')
    await session2.shouldHaveReply('.foo bar', 'foobar')
  })
})

describe('nickname prefix', () => {
  beforeAll(() => {
    app.options.similarityCoefficient = 0
    app.options.commandPrefix = '-'
    app.prepare()
  })

  afterAll(() => {
    app.options.similarityCoefficient = 0.4
    app.options.commandPrefix = null
    app.prepare()
  })

  test('no nickname', async () => {
    await session1.shouldHaveReply('foo bar', 'foobar')
    await session2.shouldHaveNoReply('foo bar')
    await session1.shouldHaveReply('-foo bar', 'foobar')
    await session2.shouldHaveReply('-foo bar', 'foobar')
    await session2.shouldHaveReply(`[CQ:at,qq=${app.selfId}] foo bar`, 'foobar')
  })

  test('single nickname', async () => {
    app.options.nickname = 'koishi'
    app.prepare()

    await session1.shouldHaveReply('koishi, foo bar', 'foobar')
    await session2.shouldHaveReply('koishi, foo bar', 'foobar')
    await session1.shouldHaveReply('koishi\n foo bar', 'foobar')
    await session2.shouldHaveReply('koishi\n foo bar', 'foobar')
    await session1.shouldHaveReply('@koishi foo bar', 'foobar')
    await session2.shouldHaveReply('@koishi foo bar', 'foobar')
    await session1.shouldHaveNoReply('komeiji, foo bar')
    await session2.shouldHaveNoReply('komeiji, foo bar')
  })

  test('multiple nicknames', async () => {
    app.options.nickname = ['komeiji', 'koishi']
    app.prepare()

    await session1.shouldHaveReply('foo bar', 'foobar')
    await session2.shouldHaveNoReply('foo bar')
    await session1.shouldHaveReply('-foo bar', 'foobar')
    await session2.shouldHaveReply('-foo bar', 'foobar')
    await session1.shouldHaveReply('koishi, foo bar', 'foobar')
    await session2.shouldHaveReply('koishi, foo bar', 'foobar')
    await session1.shouldHaveReply('komeiji foo bar', 'foobar')
    await session2.shouldHaveReply('komeiji foo bar', 'foobar')
  })
})

describe('Command Execution', () => {
  const meta: Meta<'message'> = {
    userId: 789,
    selfId: app.selfId,
    postType: 'message',
    messageType: 'private',
  }

  // make coverage happy
  // equal to: app.server.parseMeta(meta)
  app.executeCommandLine('', meta)
  const send = meta.$send = jest.fn()

  beforeEach(() => send.mockClear())

  test('app.executeCommandLine (nonexistent)', async () => {
    const next = jest.fn()
    await app.executeCommandLine('no-such-command', meta, next)
    expect(send).toBeCalledTimes(0)
    expect(next).toBeCalledTimes(1)
  })

  test('context.runCommand (nonexistent)', async () => {
    await app.runCommand('no-such-command', meta)
    expect(send).toBeCalledWith(messages.COMMAND_NOT_FOUND)
  })

  test('command error', async () => {
    const error = new Error('command error')
    app.command('error-command').action(() => { throw error })
    const errorCallback = jest.fn()
    const errorCommandCallback = jest.fn()
    app.on('error', errorCallback)
    app.on('error/command', errorCommandCallback)
    await app.executeCommandLine('error-command', meta)
    expect(errorCallback).toBeCalledTimes(1)
    expect(errorCallback).toBeCalledWith(error)
    expect(errorCommandCallback).toBeCalledTimes(1)
    expect(errorCommandCallback).toBeCalledWith(error)
  })

  test('command events', async () => {
    app.command('skipped-command').action(({ next }) => next())
    const beforeCommandCallback = jest.fn()
    const afterCommandCallback = jest.fn()
    app.on('before-command', beforeCommandCallback)
    app.on('after-command', afterCommandCallback)
    await app.executeCommandLine('skipped-command', meta)
    expect(beforeCommandCallback).toBeCalledTimes(1)
    expect(afterCommandCallback).toBeCalledTimes(0)
  })

  test('insufficient arguments', async () => {
    await app.runCommand('foo', meta)
    expect(send).toBeCalledWith(messages.INSUFFICIENT_ARGUMENTS)
  })

  test('redundant arguments', async () => {
    await app.runCommand('foo', meta, ['bar', 'baz'])
    expect(send).toBeCalledWith(messages.REDUNANT_ARGUMENTS)
  })

  test('required options', async () => {
    await app.runCommand('fooo', meta)
    expect(send).toBeCalledWith(format(messages.REQUIRED_OPTIONS, '-t, --text <bar>'))
  })

  test('unknown options', async () => {
    await app.runCommand('fooo', meta, [], { text: 'bar', test: 'baz' })
    expect(send).toBeCalledWith(format(messages.UNKNOWN_OPTIONS, 'test'))
  })
})

describe('shortcuts', () => {
  beforeAll(() => {
    app.options.commandPrefix = '#'
    app.prepare()
  })

  afterAll(() => {
    app.options.commandPrefix = null
    app.prepare()
  })

  test('single shortcut', async () => {
    await session2.shouldHaveReply(' bar1 ', 'foobar')
    await session2.shouldHaveReply(' bar2 ', 'fooobar')
    await session2.shouldHaveNoReply('bar1 bar')
    await session2.shouldHaveNoReply('bar2 -t bar')
  })

  test('no command prefix', async () => {
    await session2.shouldHaveNoReply('#bar1')
    await session2.shouldHaveNoReply('#bar2')
    await session2.shouldHaveNoReply('#bar3')
    await session2.shouldHaveNoReply('#baz')
  })

  test('nickname prefix & fuzzy', async () => {
    await session2.shouldHaveNoReply('bar3 -t baz')
    await session2.shouldHaveReply(`[CQ:at,qq=${app.selfId}] bar3 -t baz`, 'fooobaz')
  })

  test('one argument & fuzzy', async () => {
    await session2.shouldHaveReply('bar4 bar baz', 'foobar baz')
    await session2.shouldHaveNoReply('bar4bar baz')
    await session2.shouldHaveReply(`[CQ:at,qq=${app.selfId}] bar4bar baz`, 'foobar baz')
  })
})
