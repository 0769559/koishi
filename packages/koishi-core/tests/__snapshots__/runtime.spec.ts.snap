// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`suggestions apply suggestions:   1`] = `"foobar"`;

exports[`suggestions apply suggestions: fo bar 1`] = `"没有此命令。你要找的是不是“foo”？发送空行以调用推测的指令。"`;

exports[`suggestions execute command: foo bar 1`] = `"foobar"`;

exports[`suggestions ignore suggestions 1: fo bar 1`] = `"没有此命令。你要找的是不是“foo”？发送空行以调用推测的指令。"`;

exports[`suggestions ignore suggestions 2: fo bar 1`] = `"没有此命令。你要找的是不是“foo”？发送空行以调用推测的指令。"`;

exports[`suggestions multiple suggestions: fool bar 1`] = `"没有此命令。你要找的是不是“foo”或“fooo”？"`;
